irohad/ametsuchi/impl/storage_impl.cpp

```diff
@@ -687,78 +687,101 @@ namespace iroha {
       return storage;
     }
 
-    boost::optional<std::unique_ptr<LedgerState>> StorageImpl::commit(
+    CommitResult StorageImpl::commit(
         std::unique_ptr<MutableStorage> mutable_storage) {
       auto storage = static_cast<MutableStorageImpl *>(mutable_storage.get());
 
       try {
         *(storage->sql_) << "COMMIT";
-        storage->committed = true;
-
-        storage->block_storage_->forEach(
-            [this](const auto &block) { this->storeBlock(block); });
-
-        return PostgresWsvQuery(*(storage->sql_),
-                                factory_,
-                                log_manager_->getChild("WsvQuery")->getLogger())
-                   .getPeers()
-            | [&storage](auto &&peers) {
-                return boost::optional<std::unique_ptr<LedgerState>>(
-                    std::make_unique<LedgerState>(std::move(peers),
-                                                  storage->getTopBlockHeight(),
-                                                  storage->getTopBlockHash()));
-              };
       } catch (std::exception &e) {
         storage->committed = false;
         log_->warn("Mutable storage is not committed. Reason: {}", e.what());
-        return boost::none;
+        return expected::makeError(e.what());
+      }
+      storage->committed = true;
+
+      shared_model::interface::types::HeightType top_height;
+      shared_model::crypto::Hash top_hash;
+      storage->block_storage_->forEach(
+          [this, &top_height, &top_hash](const auto &block) {
+            this->storeBlock(block);
+            top_height = block->height();
+            top_hash = block->hash();
+          });
+
+      decltype(std::declval<PostgresWsvQuery>().getPeers()) opt_ledger_peers;
+      {
+        auto peer_query =
+            PostgresWsvQuery(*(storage->sql_),
+                             factory_,
+                             log_manager_->getChild("WsvQuery")->getLogger());
+        while (not(opt_ledger_peers = peer_query.getPeers())) {
+          log_->error("Failed to get ledger peers! Will retry.");
+        }
       }
+      assert(opt_ledger_peers);
+
+      return expected::makeValue(std::make_shared<LedgerState>(
+          std::move(*opt_ledger_peers), top_height, top_hash));
     }
 
-    boost::optional<std::unique_ptr<LedgerState>> StorageImpl::commitPrepared(
+    boost::optional<CommitResult> StorageImpl::commitPrepared(
         std::shared_ptr<const shared_model::interface::Block> block) {
       if (not prepared_blocks_enabled_) {
         log_->warn("prepared blocks are not enabled");
         return boost::none;
```

Comment:
This is needed because we use this function to test whether prepared blocks are enabled. When the commit does not happen due to prepared commits being disabled - it is not an error, so it would be better to distinguish such situation from a real error.
But we can also make a dedicated method to check whether prepared txs are enabled, and then treat a call to `commitPrepared` when they are disabled like a real error.